connectors/sources/sharepoint_online.py

```diff
@@ -0,0 +1,740 @@
+            # Yes, makes no sense to me either.
+            return
+
+    async def download_attachment(self, attachment_absolute_path, async_buffer):
+        self._validate_sharepoint_rest_url(attachment_absolute_path)
+
+        await self._rest_api_client.pipe(
+            f"{attachment_absolute_path}/$value", async_buffer
+        )
+
+    async def site_pages(self, site_web_url):
+        self._validate_sharepoint_rest_url(site_web_url)
+
+        select = ""
+        url = f"{site_web_url}/_api/web/lists/GetByTitle('Site%20Pages')/items?$select={select}"
+
+        async for page in self._rest_api_client.scroll(url):
+            for site_page in page:
+                yield site_page
+
+    def _validate_sharepoint_rest_url(self, url):
+        if "OVERRIDE_URL" in os.environ:  # TODO: make it in a call instead of this
+            return
+
+        # I haven't found a better way to validate tenant name for now.
+        actual_tenant_name = self._tenant_name_pattern.findall(url)[0]
+
+        if self._tenant_name != actual_tenant_name:
+            raise Exception(
+                f"Unable to call Sharepoint REST API - tenant name is invalid. Authenticated for tenant name: {self._tenant_name}, actual tenant name for the service: {actual_tenant_name}."
+            )
+
+    async def close(self):
+        await self._http_session.close()
+
+
+class SharepointOnlineAdvancedRulesValidator(AdvancedRulesValidator):
+    """
+    Validate advanced rules for MongoDB, so that they're adhering to the motor asyncio API (see: https://motor.readthedocs.io/en/stable/api-asyncio/asyncio_motor_collection.html)
+    """
+
+    SCHEMA_DEFINITION = {
+        "type": "object",
+        "properties": {
+            "maxDataAge": {"type": "integer"},
+        },
+        "additionalProperties": False,
+    }
+
+    SCHEMA = fastjsonschema.compile(definition=SCHEMA_DEFINITION)
+
+    async def validate(self, advanced_rules):
+        try:
+            SharepointOnlineAdvancedRulesValidator.SCHEMA(advanced_rules)
+
+            return SyncRuleValidationResult.valid_result(
+                rule_id=SyncRuleValidationResult.ADVANCED_RULES
+            )
+        except JsonSchemaValueException as e:
+            return SyncRuleValidationResult(
+                rule_id=SyncRuleValidationResult.ADVANCED_RULES,
+                is_valid=False,
+                validation_message=f"{e.message}. Make sure advanced filtering rules follow the following schema: {SharepointOnlineAdvancedRulesValidator.SCHEMA_DEFINITION['properties']}",
+            )
+
+
+class SharepointOnlineDataSource(BaseDataSource):
+    """Sharepoint Online"""
+
+    name = "Sharepoint Online"
+    service_type = "sharepoint_online"
+
+    def __init__(self, configuration):
+        super().__init__(configuration=configuration)
+
+        self._client = None
+
+    @property
+    def client(self):
```

Comment:
You can use [@cache](https://docs.python.org/3/library/functools.html#functools.cache)

see example: 
https://github.com/elastic/connectors-python/blob/7539174dfb0a35a1129e0b992e3c5702080b7e98/connectors/source.py#L588

Reply:
Thanks, will do!